app.py

```diff
@@ -1,31 +1,307 @@
+    country = SelectField('Country of Origin', choices=country_list , validators=[validators.DataRequired()])
+
+class Username(Form):
+    #Set up form
+    username = TextField('Username:', validators=[validators.DataRequired("*Required")])
+
+class Search(Form):
+    #Set up form
+    search = TextField('Search:', validators=[validators.DataRequired("*Required")])
+
+@app.route('/', methods=['GET','POST'])
+def index():
+    #Username form
+    wtform = Username(request.form)
+    if wtform.validate():
+        return redirect('/' + request.form['username'] + '/recipes?limit=10&offset=0')
+    
+    return render_template("index.html", form=wtform, errors=wtform.errors)
+    
+@app.route('/<username>/recipes')
+def recipes(username):
+    
+    # Get All Recipes
+    recipes = mongo.db.recipes
+    all_recipes = recipes.find()
+    
+    # Pagination Settings
+    offset = 0 
+    limit = int(request.args.get('limit'))
+    
+    #Get Count
+    count_list = []
+    for doc in all_recipes:
+        count_list.append(doc)
+        count = len(count_list)
+    
+    #Get Pages And Generate URL List
+    pages = get_pages(count, limit)
+    url_list = generate_pagination_links(offset, limit, pages, 'recipes', 'null', username)
+    print(url_list)
+    
+    #Get _id of Last Item on a Page
+    dynamic_position = request.args.get('offset')
+    starting_id = recipes.find().sort('_id')
+    last_id = starting_id[int(dynamic_position)]['_id']
+    
+    #Sort Tables
+    sort_default = recipes.find({'_id' : {'$gte' : last_id}}).limit(limit)
+    sort_country = recipes.find({'_id' : {'$gte' : last_id}}).sort([( "country", 1),
+    ("name", 1 )]).limit(limit)
+    sort_name = recipes.find({'_id' : {'$gte' : last_id}}).sort([("name",
+    1 )]).limit(limit)
+    sort_upvotes = recipes.find({'_id' : {'$gte' : last_id}}).sort([( "upvotes",
+    pymongo.DESCENDING), ("name",1 )]).limit(limit)
+    sort_downvotes = recipes.find({'_id' : {'$gte' : last_id}}).sort([( "downvotes",pymongo.DESCENDING), ("name", 1 )]).limit(limit)
+    sort_author = recipes.find({'_id' : {'$gte' : last_id}}).sort([( "author",
+    1), ("name", 1 )]).limit(limit)
+   
+    return render_template("recipes.html", author=sort_author, 
+    default=sort_default, name=sort_name, upvotes=sort_upvotes, 
+    downvotes= sort_downvotes, country=sort_country, url_list=url_list, 
+    pages=pages, username=username)
+
+@app.route('/<username>/search', methods=['GET','POST'])
+def search(username):
+    #Search User Input
+        wtform = Search(request.form)
+        if wtform.validate():
+            return redirect('/' + username + '/' + 'search' + '/' + request.form["search"] + '?limit=10&offset=0')
+        return render_template("search.html", username=username, form=wtform, errors=wtform.errors )
+    
+@app.route('/<username>/search/<search>', methods=['GET','POST'] )
+def results(username, search):
+    
+    
+    # Get All Recipes
+    recipes = mongo.db.recipes
```

Comment:
Same as above... use controllers.py. Ensure not to repeat this statement. Create another file to import mongo and return `recipes` then import same to your controllers.py